### ISSUE TITLE
iOS app hangs when processing lots of alarm notifications at login

### ISSUE THREAD

**Issue by charlag** [tutao/tutanota#3010]
Edited 27.07.21

The iOS app will freeze and potentially crash at login if the are a lot of alarm notifications
to reproduce:
* login to an account that has a large calendar, you might need to prepare this by importing a *big* ics file from the web client
* the app will hang for a while, and possibly crash

It looks like there are at least 2 problems:
* there seems to be a memory leak somewhere in `TUTAlarmManager.handleAlarmNotification` which it being called during a loop in `processNewAlarms`. Wrapping the call in an `@autoreleasepool` block seems to fix it, i don't know enough about iOS to know if this is a good fix.
* The implementation looks like it could be optimized - we read and write from device config storage for every alarm, where we could just be doing one write after processing is done, so it takes a very long time when there are alot of events. If we can't speed up the implementation significantly, we should make it happen concurrently so as not to hang the UI.

---

Test notes:
 - [ ] Force stop iOS app. Import big file with alarms. Start iOS app. See that it works normally and alarms are scheduled.

**Comment by IAteYourBacon** [comment]
Here is how to reproduce it:
When I delete the app and the data (through Settings > Storage) and reinstall it. it works, I can access the mailbox. But if I double click the home button and swipe up on tutanota to close it while being logged in, then at the next start the app will pass the log in phase but freeze on loading the message list and crash. It will never load the message list again until I delete app and data and reinstall again.
If I launch the app after a fresh install but don't log in and close it with that home button double click and swipe up, then I'm able to open the app again, log in and load the message list.
The crash happens only when I'm logged in and close the app.
iPhone 6
IOS : 12.5.3
Tutanota app : 3.82.19

In the logs I included, it shows "Hardware Model:      iPhone7,2" but it's actually an iphone 6, I don't know if it's relevant.
[tutanota-2021-05-26-072846.ips.tar.gz](https://github.com/tutao/tutanota/files/6554154/tutanota-2021-05-26-072846.ips.tar.gz)

**Comment by charlag** [comment]
@IAteYourBacon thank you for the info, I guess I got your mail in support!

Here's the reason:

```
Termination Reason: Namespace SPRINGBOARD, Code 0x8badf00d
Termination Description: SPRINGBOARD, process-exit watchdog transgression: de.tutao.tutanota exhausted real (wall clock) time allowance of 5.00 seconds | ProcessVisibility: Background | ProcessState: Running | WatchdogEvent: process-exit | WatchdogVisibility: Background | WatchdogCPUStatistics: ( | "Elapsed total CPU time (seconds): 9.680 (user 9.680, system 0.000), 97% CPU", | "Elapsed application CPU time (seconds): 1.253, 13% CPU" | )
Triggered by Thread:  0
```

so it seems to freeze. It could be some alarm calculation or something else that blocks main thread.

**Comment by IAteYourBacon** [comment]
@charlag You were right, after deleting all the calendars, no more crash. I can quit the app with a home button double click and swipe up, then reopen it and access the mailbox without crashing it.

**Comment by charlag** [comment]
@IAteYourBacon thanks a lot for the info! We now know in which direction to dig.

**Comment by IAteYourBacon** [comment]
ok, so it seems to be linked to either the quantity of events in a calendar or the import of a .ics file . 
I added shared calendars : no crash
I added a repeating event with reminders to the default local calendar : no crash
I added an event repeating 50 times with reminders to the a second local calendar : no crash
After importing a 631KB .ics file from a google calendar export : the app crashes again

The import was done in tutanota desktop on an intel mac and took about 20mn, I have had about 20 notifications saying "Failed to handle PushMessage" during the import. Maybe that's where something is going wrong.

I hope that helps

**Comment by McLapngap** [comment]
> ok, so it seems to be linked to either the quantity of events in a calendar or the import of a .ics file .
> I added shared calendars : no crash
> I added a repeating event with reminders to the default local calendar : no crash
> I added an event repeating 50 times with reminders to the a second local calendar : no crash
> After importing a 631KB .ics file from a google calendar export : the app crashes again
> 
> The import was done in tutanota desktop on an intel mac and took about 20mn, I have had about 20 notifications saying "Failed to handle PushMessage" during the import. Maybe that's where something is going wrong.
> 
> I hope that helps

Thank you for sharing this. I seem to have found a match of what I am experiencing! In my case the hang and freeze with the IOS (14.6) app started after importing a long ics file using a desktop web browser. It freezes within 1 minute of opening the app, either when opening an email, scrolling down, or go to a setting page...

I deleted the imported calendar from Tutanota and hang seems to have gone away for now.

Looking forward to a fix in the next release and thanks in advance.

**Comment by johnbotris** [comment]
still freezes and crashes

**Comment by johnbotris** [comment]
closed by 8088a7a679e1a48e6ca286580383ad6228e03b61
